### PR TITLE
Skeletal Resource: Aws Vpc Subnets

### PR DESCRIPTION
**Pull request by dromazmj**
* Adds functionality for testing all or a group of subnets

### CODE REVIEW

File: docs/resources/aws_vpc_subnets.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+<br>
+
+## Syntax
+
+An `aws_vpc_subnets` resource block uses an optional filter to select a group of subnets and then tests that group.
+
+    # Test all subnets within a single vpc
+    describe aws_vpc_subnets.where(vpc_id: 'vpc-12345678') do
+      its('subnet_id') { should include ['subnet-12345678', 'subnet-98765432'] }
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: I don't think `include` takes an array, typically.  You'd want to say 
```
its('subnet_id') { should include 'subnet-12345678' }
its('subnet_id') { should include 'subnet-98765432' }
```

File: docs/resources/aws_vpc_subnets.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    describe aws_vpc_subnets.where( vpc_id: 'vpc-12345678') do
+      its('subnet-ids') { should include ['subnet-12345678', 'subnet-98765432'] }
+    end
+
+### subnet_id
+
+A string identifying a specific subnet.
+
+    # Examine a specific subnet
+    describe aws_ec2_security_groups.where(subnet_id: 'subnet-12345678') do
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: Copypasta.  The resource is named aws_vpc_subnets.

File: docs/resources/aws_vpc_subnets.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+## Filter Criteria
+
+### vpc_id
+
+A string identifying the VPC which may or may not contain subnets.
+
+    # Look for all subnts within a vpc.
+    describe aws_vpc_subnets.where( vpc_id: 'vpc-12345678') do
+      its('subnet-ids') { should include ['subnet-12345678', 'subnet-98765432'] }
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: The property name in the example has a dash, while the property (below) has an underscore.

File: docs/resources/aws_vpc_subnets.md (hunk 1 above)
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: Below, the property is plural (which I think reads better) - `subnet_ids`, not `subnet_id`.

File: docs/resources/aws_vpc_subnets.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+## Properties
+
+### cidr_blocks
+
+Provides a string that contains the cidr block of ip addresses that can be given in the subnet.
+
+    # Examine a specific subnets cidr_block
+    describe aws_ec2_security_groups.where( subnet_id: 'subnet-12345678') do
+      its('cidr_block') { should eq ['10.0.1.0/24'] }
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: Property name should be plural in the example.

File: docs/resources/aws_vpc_subnets.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    end
+
+### vpc_ids
+
+Provides an array containing a string of the vpc_id associated with a subnet.
+
+    # Examine a specific subnets cidr_block
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: Examine VPC ID.  Copypasta.

File: docs/resources/aws_vpc_subnets.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    describe aws_ec2_security_groups.where( subnet_id: 'subnet-12345678') do
+      its('vpc_ids') { should eq ['vpc-12345678'] }
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: include 'vpc-12345678', not eq [ ... ]

File: docs/resources/aws_vpc_subnets.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    end
+
+### subnet_ids
+
+Provides an array of strings containing the subnet IDs associated with a vpc
+
+    # Examine a specific subnets cidr_block
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: Copypasta.  Not cidr_block, subnet_ids.

File: docs/resources/aws_vpc_subnets.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    describe aws_ec2_security_groups.where( vpc_id: 'vpc-12345678') do
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: Copypasta.  Incoorrect resource name.

File: docs/resources/aws_vpc_subnets.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+      its('subnet_ids') { should eq ['subnet-12345678', 'subnet-87654321'] }
```
Comment (clintoncwolfe) [inline docs/resources/aws_vpc_subnets.md]: Use include, not eq-array.

File: libraries/aws_vpc_subnets.rb [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+class AwsVpcSubnets < Inspec.resource(1)
+  name 'aws_vpc_subnets'
+  desc 'Verifies settings for VPC Subnets in bulk'
+  example "
+    # you should be able to test the cidr_block of a subnet
+    describe aws_vpc_subnets.where(vpc_id: 'vpc-123456789') do
+      its('subnet_ids') { should eq ['subnet-12345678', 'subnet-87654321'] }
+      its('cidr_blocks') { should eq ['172.31.96.0/20'] }
+      its('states') { should include 'available' }
```
Comment (clintoncwolfe) [inline libraries/aws_vpc_subnets.rb]: `states` is mentioned here, but  not in the Markdown docs.

File: libraries/aws_vpc_subnets.rb [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+    # No criteria yet
+    recognized_criteria = check_criteria_names(raw_criteria)
+    recognized_criteria
+  end
+
+  def check_criteria_names(raw_criteria: {}, allowed_criteria: [])
+    # Remove all expected criteria from the raw criteria hash
+    recognized_criteria = {}
+    allowed_criteria.each do |expected_criterion|
```
Comment (clintoncwolfe) [inline libraries/aws_vpc_subnets.rb]: Where is allowed_criteria set?  Was this tested?

File: libraries/aws_vpc_subnets.rb [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+      raise 'Unrecognized criteria for fetching Vpc Subnets. ' \
+            "Use 'criteria: value' format."
+    end
+
+    # No criteria yet
+    recognized_criteria = check_criteria_names(raw_criteria)
+    recognized_criteria
+  end
+
+  def check_criteria_names(raw_criteria: {}, allowed_criteria: [])
```
Comment (clintoncwolfe) [inline libraries/aws_vpc_subnets.rb]: If you're not accepting any criteria, this whole method is unneeded.

File: test/integration/default/verify/controls/aws_vpc_subnets.rb [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  'ec2_default_vpc_subnet_id',
+].each do |fixture_name|
+  fixtures[fixture_name] = attribute(
+    fixture_name,
+    default: "default.#{fixture_name}",
+    description: 'See ../build/ec2.tf',
+  )
+end
+
+control "aws_vpc_subnets client side filtering" do
```
Comment (clintoncwolfe) [inline test/integration/default/verify/controls/aws_vpc_subnets.rb]: This resource has no server-side filtering, so there is no distinction between client and server-side filtering.  Puzzled at the control name.

File: test/integration/default/verify/controls/aws_vpc_subnets.rb [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+  describe all_subnets.where(subnet_id: 'subnet-00000000') do
+    it { should_not exist }
+  end
+end
+
+control "aws_vpc_subnets properties of default VPC subnet" do
+  # you should be able to test the cidr_block of a subnet
+  describe aws_vpc_subnets.where(subnet_id: fixtures['ec2_default_vpc_subnet_id']) do
+    its('cidr_blocks') { should eq ['172.31.96.0/20'] }
+    its('states') { should eq ['available'] }
```
Comment (clintoncwolfe) [inline test/integration/default/verify/controls/aws_vpc_subnets.rb]: Change to include-string, not eq-array.  include-string is much easier for non-developers to read, and auditors / security perssonel often look to our integration tests for examples.

File: test/integration/default/verify/controls/aws_vpc_subnets.rb [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+  end
+end
+
+control "aws_vpc_subnets properties of default VPC" do
+  # you should be able to test the cidr_block of a subnet
+  describe aws_vpc_subnets.where(vpc_id: fixtures['ec2_security_group_default_vpc_id']) do
+    its('cidr_blocks') { should include '172.31.96.0/20' }
+    its('states') { should include 'available' }
```
Comment (clintoncwolfe) [inline test/integration/default/verify/controls/aws_vpc_subnets.rb]: 💯

File: test/unit/resources/aws_vpc_subnets_test.rb [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+        }),
+      ]
+
+      selected = fixtures.select do |sg|
+        query.keys.all? do |criterion|
+          query[criterion] == sg[criterion]
+        end
+      end
+
+      OpenStruct.new({ subnets: selected })
```
Comment (clintoncwolfe) [inline test/unit/resources/aws_vpc_subnets_test.rb]: What happens if the AWS SDK can't find the requested subnet?  If it throws an exception, the unit test fixture should too.

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
Overall very good.  Most issues in the docs or examples.  Very solid tests!  Thanks!

File: test/unit/resources/aws_vpc_subnets_test.rb (hunk 16 above)
Comment (dromazmj) [reply test/unit/resources/aws_vpc_subnets_test.rb]: It returns an empty

**Review by clintoncwolfe — APPROVED** [review]

**Comment by clintoncwolfe** [comment]
This PR looks great, but has a failing unit test with an obscure error, due to the FilterTable data being an array of OpenStructs, rather than an array of hashes.  I've adopted the PR as #228, and added a fix there.  Many thanks!

**Pull request closed without merging**